README.md
# Carousel Select Widget

An iOS style alternative to dropdown value picker with horizontal and vertical scroll mode.

## Getting Started

To use this package, add `carousel_select_widget` as a [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages)

## Example

Import the library.

```
import 'package:carousel_select_widget/carousel_select_widget.dart';
```

Then invoke the `CarouselSelect` constructor with appropriate properties to display the widget.

## Sample Code

```
import 'package:flutter/material.dart';
import 'package:carousel_select_widget/carousel_select_widget.dart';

void main() => runApp(CarouselSelectWidgetSample());

const List<String> kCurrenciesList = ['AUD', 'BRL', 'CAD', 'CNY', 'EUR', 'GBP', 'HKD', 'IDR', 'ILS', 'INR'];
const int kInitialPositionIndex = 0;
const Color kDarkBackgroundColor = Color(0xFF111428);
const Color kTurquoiseColor = Color(0xFF1abc9c);
const Color kMidnightBlue = Color(0xFF2c3e50);
const Color kActiveItemTextColor = Color(0xFF9b59b6);
const double kActiveItemFontSize = 20;
const double kPassiveItemFontSize = 20;
const double kHorizontalWidgetSize = 100.0;
const double kVerticalWidgetSize = 200.0;

class CarouselSelectWidgetSample extends StatefulWidget {
  @override
  _CarouselSelectWidgetSampleState createState() => _CarouselSelectWidgetSampleState();
}

class _CarouselSelectWidgetSampleState extends State<CarouselSelectWidgetSample> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        backgroundColor: kDarkBackgroundColor,
        appBar: AppBar(
          title: Center(
            child: Text('Carousel Select Widget Example',)
            ),
          backgroundColor: kDarkBackgroundColor,
          elevation: 0,
        ),
        body: Center(
          child: Container(
            child: CarouselSelect(
              onChanged: (selectedValue) {print(selectedValue);},
              valueList: kCurrenciesList,
              backgroundColor: kDarkBackgroundColor,
              activeItemTextColor: kActiveItemTextColor,
              passiveItemsTextColor: kActiveItemTextColor.withOpacity(0.3),
              initialPosition: kInitialPositionIndex,
              scrollDirection: ScrollDirection.horizontal,
              activeItemFontSize: kActiveItemFontSize,
              passiveItemFontSize: kPassiveItemFontSize,
              height: kHorizontalWidgetSize,
              ),
            ),
          ),
        ),
      );
    }
}
```

## Sample App

Refer to example folder.

## Known Issue

* Using both horizontal and vertical scroll direction on the same page causes glitch in horizontal widget.

## Credits

This minimal version of widget is inspired from beautifully designed 'horizontal_picker' widget by [kaiserleka](https://pub.dev/packages/horizontal_picker)

CHANGELOG.md
## [0.0.1] - 08 January 2020

* Inital release


lib/carousel_select_widget.dart
library carousel_select_widget;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

enum ScrollDirection {horizontal, vertical}
int _listRotation;
int _listItemRotation;

class CarouselSelect extends StatefulWidget {
  // properties
  final Function(String) onChanged;
  final int initialPosition;
  final ScrollDirection scrollDirection;
  final Color backgroundColor;
  final Color activeItemTextColor;
  final Color passiveItemsTextColor;
  final double activeItemFontSize;
  final double passiveItemFontSize;
  final List<String> valueList;
  final double height;

  // contructor
  CarouselSelect(
      {@required this.valueList,
        @required this.initialPosition,
        @required this.scrollDirection,
        @required this.onChanged,
        this.backgroundColor = Colors.white,
        this.activeItemTextColor = Colors.blue,
        this.passiveItemsTextColor = Colors.grey,
        this.activeItemFontSize = 20.0,
        this.passiveItemFontSize = 20.0,
        this.height = 100.0})
      : assert(onChanged != null);

  @override
  _CarouselSelectState createState() => _CarouselSelectState();
}

class _CarouselSelectState extends State<CarouselSelect> {
  // similar to scrollcontroller but with added mechanism to stop and read the given indices
  FixedExtentScrollController _scrollController;

  // to store value, fontsize, and color of the text in the picker
  List<Map> valueMap = [];

  @override
  void initState() {
    super.initState();

    for (var i = 0; i <= widget.valueList.length - 1; i++) {
      valueMap.add({
        "value": widget.valueList[i],
        "fontSize": widget.passiveItemFontSize,
        "color": widget.passiveItemsTextColor,
      });
    }
    setScrollController();
  }

  setScrollController() {
    //sets initial scroll positon to center
    int initialItem = widget.initialPosition;

    // sets the selector to initial posti
    _scrollController = FixedExtentScrollController(initialItem: initialItem);

    // applies active style to initial selected item
    valueMap[initialItem]["color"] = widget.activeItemTextColor;
    valueMap[initialItem]["fontSize"] = widget.activeItemFontSize;
  }


  @override
  Widget build(BuildContext context) {
    int curItem;

    switch (widget.scrollDirection) {
      case ScrollDirection.horizontal:
        _listRotation = 3;
        _listItemRotation = 1;
        break;
      case ScrollDirection.vertical:
        _listRotation = 2;
        _listItemRotation = 2;
        break;
    }

    return Container(
      height: widget.height,
      width: double.infinity,
      alignment: Alignment.center,
      child: Scaffold(
        backgroundColor: widget.backgroundColor,
        body: RotatedBox(
          quarterTurns: _listRotation,
          child: ListWheelScrollView(
              controller: _scrollController,
              itemExtent: 60,
              onSelectedItemChanged: (item) {
                curItem = item;
                setState(() {
                  widget.onChanged(valueMap[curItem]["value"]);
                  for (var i = 0; i < valueMap.length; i++) {
                    if (i == curItem) {
                      valueMap[curItem]["color"] = widget.activeItemTextColor;
                      valueMap[curItem]["fontSize"] = widget.activeItemFontSize;
                    } else {
                      valueMap[i]["color"] = widget.passiveItemsTextColor;
                      valueMap[i]["fontSize"] = widget.passiveItemFontSize;
                    }
                  }
                });
              },
              children: valueMap.map((Map curValue) {
                return ItemWidget(
                  curValue,
                  widget.backgroundColor,
                );
              }).toList()),
        ),
      ),
    );
  }
}

class ItemWidget extends StatefulWidget {

  final Map curItem;
  final Color backgroundColor;
  ItemWidget(this.curItem, this.backgroundColor);

  @override
  _ItemWidgetState createState() => _ItemWidgetState();
}

class _ItemWidgetState extends State<ItemWidget> {
  String textParts;
  @override
  void initState() {
    super.initState();
    var mtext = widget.curItem["value"];
    textParts = mtext;
  }

  @override
  Widget build(BuildContext context) {
    return RotatedBox(
      quarterTurns: _listItemRotation,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Text(
            textParts,
            style: TextStyle(
                fontSize: widget.curItem["fontSize"],
                color: widget.curItem["color"]),
            textAlign: TextAlign.center,
            softWrap: false,
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:carousel_select_widget/carousel_select_widget.dart';

const List<String> kCurrenciesList = [
  'AUD',
  'BRL',
  'CAD',
  'CNY',
  'EUR',
  'GBP',
  'HKD',
  'IDR',
  'ILS',
  'INR',
  'JPY',
  'MXN',
  'NOK',
  'NZD',
  'PLN',
  'RON',
  'RUB',
  'SEK',
  'SGD',
  'USD',
  'ZAR'
];

const List<String> kCryptoList = [
  'BTC',
  'ETH',
  'LTC',
];

const int kInitialPositionIndex = 0;
const Color kDarkBackgroundColor = Color(0xFF111428);
const Color kTurquoiseColor = Color(0xFF1abc9c);
const Color kMidnightBlue = Color(0xFF2c3e50);
const Color kActiveItemTextColor = Color(0xFF9b59b6);
const double kActiveItemFontSize = 20;
const double kPassiveItemFontSize = 20;
const double kHorizontalWidgetSize = 100.0;
const double kVerticalWidgetSize = 200.0;
const String kFontFamily = 'Montserrat';

void main() => runApp(CarouselSelectWidgetExample());

class CarouselSelectWidgetExample extends StatefulWidget {
  @override
  _CarouselSelectWidgetExampleState createState() =>
      _CarouselSelectWidgetExampleState();
}

class _CarouselSelectWidgetExampleState
    extends State<CarouselSelectWidgetExample> {
  
  String _hselectedCurrency = kCurrenciesList[kInitialPositionIndex];
  String _vselectedCurrency = kCryptoList[kInitialPositionIndex];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(brightness: Brightness.dark, fontFamily: kFontFamily),
      home: Scaffold(
        backgroundColor: kDarkBackgroundColor,
        appBar: AppBar(
          title: Center(child: Text('Carousel Select Widget Example',)),
          backgroundColor: kDarkBackgroundColor,
          elevation: 0,
        ),
        body: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: <Widget>[
                Text(_hselectedCurrency, style: TextStyle(fontSize: 50, color: kTurquoiseColor),),
                Text(_vselectedCurrency, style: TextStyle(fontSize: 50, color: kTurquoiseColor),),
              ],
            ),
            Divider(indent: 50.0, endIndent: 50.0, color: kMidnightBlue,),
            Text('HORIZONTAL CAROUSEL SELECT'),
            CarouselSelect(
              onChanged: (value) {
                setState(() {
                  _hselectedCurrency = value;
                });
              },
              valueList: kCurrenciesList,
              backgroundColor: kDarkBackgroundColor,
              activeItemTextColor: kActiveItemTextColor,
              passiveItemsTextColor: kActiveItemTextColor.withOpacity(0.3),
              initialPosition: kInitialPositionIndex,
              scrollDirection: ScrollDirection.horizontal,
              activeItemFontSize: kActiveItemFontSize,
              passiveItemFontSize: kPassiveItemFontSize,
              height: kHorizontalWidgetSize,
            ),
            Divider(indent: 50.0, endIndent: 50.0, color: kMidnightBlue,),
            Text('VERTICAL CAROUSEL SELECT'),
            CarouselSelect(
              onChanged: (value) {
                setState(() {
                  _vselectedCurrency = value;
                });
              },
              valueList: kCryptoList,
              backgroundColor: kDarkBackgroundColor,
              activeItemTextColor: kActiveItemTextColor,
              passiveItemsTextColor: kActiveItemTextColor.withOpacity(0.3),
              initialPosition: kInitialPositionIndex,
              scrollDirection: ScrollDirection.vertical,
              activeItemFontSize: kActiveItemFontSize,
              passiveItemFontSize: kPassiveItemFontSize,
              height: kVerticalWidgetSize,
            ),
            Text('CRAFTED WITH ❤️ IN FLUTTER', style: TextStyle(fontSize: 10.0), )            
          ],
        ),
      ),
    );
  }
}
